rpki-validator/src/main/java/net/ripe/rpki/validator3/domain/validation/CertificateTreeValidationService.java

```diff
@@ -291,69 +290,108 @@ private boolean isValidationRunCompleted(ValidationResult validationResult) {
-                    "rpkiObject.get", () -> rpkiObject.get(CertificateRepositoryObject.class, temporary));
-
-                if (!temporary.hasFailureForCurrentLocation()) {
-                    return maybeCertificateRepositoryObject.flatMap(certificateRepositoryObject -> {
-                        Bench.mark0(trustAnchor.getName(), "certificateRepositoryObject.validate", () ->
-                            certificateRepositoryObject.validate(location.toASCIIString(), context, crl.get(), crlUri, VALIDATION_OPTIONS, temporary));
-
-                        if (!temporary.hasFailureForCurrentLocation()) {
-                            validatedObjects.add(rpkiObject.key());
-                        }
-
-                        if (certificateRepositoryObject instanceof X509ResourceCertificate
-                            && ((X509ResourceCertificate) certificateRepositoryObject).isCa()
-                            && !temporary.hasFailureForCurrentLocation()) {
-
-                            return Optional.of(context.createChildContext(location, (X509ResourceCertificate) certificateRepositoryObject));
-                        }
-                        return Optional.empty();
-                    });
-                }
-                return Optional.<CertificateRepositoryObjectValidationContext>empty();
-            })
-                .filter(Optional::isPresent)
-                .map(Optional::get)
-                .collect(Collectors.toList())
+            manifest.getFiles().entrySet()
                 .parallelStream()
-                .map(childContext -> validateCertificateAuthority(trustAnchor, registeredRepositories, childContext, temporary))
-                .forEachOrdered(validatedObjects::addAll);
+                .map(entry -> getManifestEntry(manifestUri, entry))
+                .filter(t -> t.v2().isPresent())
+                .map(t -> t.map2(Optional::get))
+                .map(t -> getCertificateRepositoryObjectValidationContext(trustAnchor, context, validatedObjects, crlUri, x509Crl, t))
+                .filter(t -> t.v1().isPresent())
+                .map(t -> t.map1(Optional::get))
+                .map(tuple -> {
+                    final ValidationResult vr = tuple.v2();
+                    final List<Key> keys = validateCertificateAuthority(trustAnchor, registeredRepositories, tuple.v1(), vr);
+                    return new Tuple2<>(keys, vr);
+                })
+                .forEachOrdered(t -> {
+                    synchronized (validatedObjects) {
+                        validatedObjects.addAll(t.v1());
+                    }
+                    synchronized (temporary) {
+                        temporary.addAll(t.v2());
+                    }
+                });
 
         } catch (Exception e) {
-            synchronized (this) {
+            synchronized (validationResult) {
```

Comment:
Idea is unhappy synchronizing on method parameter.

Reply:
More synchronization is needed (line 231, probably at calls where it is passed in as an argument as well).

I do not think locking on the argument is feasible.